---
title: "Getting spliced and unspliced counts"
output: github_document
bibliography: bibliography.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Problem setting

We want to know quantify the the number of reads that map to introns vs exons for each gene and cell.

To do this, we need to go all the way back to the alignment (bam file), and check in the genome annotation which reads map to introns vs exons. The quantification algorithm needs to distinguish both, and produce two counts separate count matrices.

Classical pipelines, e.g. Cell Ranger, will typically ignore reads that are mapped to introns, or don't distinguish between intronic and exonic reads. However, the information is still there (as long as we did not map simply to exons).

I briefly describe two alternative pipelines here, with their advantages and disadvantages. Finally, some 

# velocyto

The velocyto pipeline was developed by the creators (Gioele La Manno, now at EPFL). It works with a python command line tool, and can be used with most analysis pipelines as long as you have an alignment (bam file) and genome annotation (gtf file). It also provides some helper functions for A typical workflow looks like this:

![](quantification.png)

http://velocyto.org/velocyto.py/tutorial/cli.html#cli

In our experience, the velocyto pipeline can take quite a while (3 hours for 5 million reads), and consume quite a bit of memory (10 GB at best). Of course, mileage may vary.

# BUStools

The BUStools pipeline is more recent, and was developed to make pseudoalignment tools useful for single-cell analysis. The central piece is kallisto, a pseudoaligner first developed for bulk RNA-seq (See http://tinyheero.github.io/2015/09/02/pseudoalignments-kallisto.html for a down-to-earth description). Pseudoalignment is much faster then exact alignment, but with a (claimed) negligible decrease in accuracy [@brayNearoptimalProbabilisticRNAseq2016;@patroSalmonFastBiasaware2017].

Central to a pseudo-aligner such as kallisto is the generation of an index based on k-mers and De Bruijn graphs. The generation of this index can take some time and memory, but can be reused between analyses. Often, only the exome is used to generate this index.

Using pseudo-aligners for single-cell analyses can be tricky because of the correct handling of UMIs. You need to keep track for every read to which gene(s) it aligned, and to which UMI this read belongs. After pseudo-alignment, the different reads of the same UMI need to be collapsed to get a single count for this UMI. For this reason (and others), the BUS format was developed, which maps cell **b**arcodes, **U**MIs and a **s**et of transcripts [@melstedBarcodeUMISet0]. In essence, the BUS file is an alternative to the BAM file, but contains less information (e.g. we don't know to where exactly a read was mapped), but is much faster to generate.

![](https://liorpachter.files.wordpress.com/2019/06/workflow.jpeg)

To use BUStools for RNA velocity, we need to generate an index which contains both introns and exons. This is the most time-consuming and memory intensive part of the whole pipeline, using over 50GB during more than an hour. Once you have the index, the rest of the process is pretty straightforward. This is nicely described in the following tutorial: https://bustools.github.io/BUS_notebooks_R/velocity.html.

# Some questions

RNA velocity analysis is still relatively new, and it is expected that the analysis pipeline will be refined a bit in the future. Here are some common issues to ponder about.

> What with reads that overlap an intron-exon junction?

These are discarded by most tools (BUStools and velocyto).

> What with alternative splicing? I.e. an exon might actually be an intron for some transcripts

These are assigned ambiguous by most tools (BUStools and velocyto) and will be discarded for downstream analyses.

> I use a 3' technology (e.g. 10X), how can I have intronic reads?!

A considerable number of reads come from "secondary priming", even when oligo-dT primers are used, as described in the original RNA velocity paper.

In some organisms (_Arabidopsis_ as an example) the number of intronic reads can be quite low, which makes RNA velocity analysis nearly impossible. One reason may be that introns are relatively short in _Arabidopsis_.